src/main/java/org/sagebionetworks/bridge/services/TemplateService.java

```diff
@@ -0,0 +1,254 @@
+        template.setGuid(generateGuid());
+        DateTime timestamp = getTimestamp();
+        template.setCreatedOn(timestamp);
+        template.setModifiedOn(timestamp);
+
+        Study study = studyService.getStudy(studyId);
+        Set<String> substudyIds = substudyService.getSubstudyIds(study.getStudyIdentifier());
+        
+        TemplateValidator validator = new TemplateValidator(study.getDataGroups(), substudyIds);
+        Validate.entityThrowingException(validator, template);
+        
+        Criteria criteria = persistCriteria(template);
+        template.setCriteria(criteria);
+        
+        templateDao.createTemplate(template);
+
+        return new GuidVersionHolder(template.getGuid(), new Long(template.getVersion()));
+    }
+    
+    public GuidVersionHolder updateTemplate(StudyIdentifier studyId, Template template) {
+        checkNotNull(studyId);
+        checkNotNull(template);
+        
+        // If the entity is deleted and staying deleted, don't allow this operation.
+        Template existing = getTemplate(studyId, template.getGuid());
+        if (existing.isDeleted() && template.isDeleted()) {
+            throw new EntityNotFoundException(Template.class);
+        }
+        checkNotDeletingDefault(template, studyId);
+        
+        template.setStudyId(studyId.getIdentifier());
+        template.setModifiedOn(getTimestamp());
+        // no reason for these to be updated after creation
+        template.setTemplateType(existing.getTemplateType());
+        template.setCreatedOn(existing.getCreatedOn());
+        
+        Study study = studyService.getStudy(studyId);
+        Set<String> substudyIds = substudyService.getSubstudyIds(study.getStudyIdentifier());
+        
+        TemplateValidator validator = new TemplateValidator(study.getDataGroups(), substudyIds);
+        Validate.entityThrowingException(validator, template);
+        
+        persistCriteria(template);
+        templateDao.updateTemplate(template);
+        
+        return new GuidVersionHolder(template.getGuid(), new Long(template.getVersion()));
+    }
+    
+    public void deleteTemplate(StudyIdentifier studyId, String guid) {
+        // This not only verifies the template exists, it verifies it is in the caller's study
+        Template existing = getTemplate(studyId, guid);
+        if (existing.isDeleted()) {
+            throw new EntityNotFoundException(Template.class);
+        }
+        checkNotDeletingDefault(existing, studyId);
+        
+        existing.setDeleted(true);
+        existing.setModifiedOn(getTimestamp());
+        
+        templateDao.updateTemplate(existing);
+    }
+    
+    public void deleteTemplatePermanently(StudyIdentifier studyId, String guid) {
+        // Deletes fail quietly, we don't throw 404s if they're unnecessary 
+        Optional<Template> optional = templateDao.getTemplate(studyId, guid);
+        if (!optional.isPresent()) {
+            return;
+        }
```

Comment:
Unit test for deleting non-existent template?

Reply:
This behavior was changed based on the design docs for logical/physical delete and tests were updated. It now throws an ENFE.